### DISCUSSION TITLE
polar form of ICtCp ?

### DISCUSSION THREAD

**Discussion by SoftMoonWebWare** [color-js/color.js#604] [Ideas]
I'm now in the process of implementing ICtCp into my MasterColorPicker project.  Adding it as a output-color-space-model is no problem.  I also want to incorporate it into a "polar" style (color-wheel) color-picker with an "intensity" slider (lightness/brightness/gray/black in other color-space-models).  I could do this also without problem, but .....
The MasterColorPicker has a "native" color-space-model-output choice, that outputs specs in whatever color-space the active color-picker itself is using. (There are 6 different graphical pickers using many different models, plus named-color tables that give you color-names when "native" is selected).
If I incorporate ICtCp into the "polar" (color-wheel) color-picker (I call it the BeezEye), what do I supply as a "native" color-space-model when it's settings are to use ICtCp ?  All the "other" rectangular spaces have polar forms (LCH, LCHuv, OKLCh) plus there is HSL, HSV, etc.  There are 12 other models the BeezEye uses, all with an angular "hue" component.  I want to keep consistency.
I'm about to call it "Ichtp", (or with my code it will be Ichᵀᴾ from ICᵀCᴾ for easier reading and instant logical comprehension; similarly I use JᶻCᶻhᶻ and Jᶻaᶻbᶻ because Jzazbz hurts my eyes trying to look at it; however, end-users will still be able to use Jᶻaᶻbᶻ or Jzazbz or ICᵀCᴾ or ICtCp)

Anyway, is there any "official" talk about this polar form of ICᵀCᴾ ?

Should I just go my own way (who me? NO! never! HaHa) and call it Ichᵀᴾ ?

(p.s. I use lowercase c for chroma here to differentiate it from the uppercase C which seems to stand for "color" in ICᵀCᴾ as in "color - tritanopia" and "color - protanopia"; and lowercase h like OKLCh does to differentiate it from a "standard RGB hue mapping" found in HSL, HSV, etc.)
Thanks again for paving the way for me!  I would be a long ways behind in development of my project, if it were not for this one.

**Comment by facelessuser** [comment]
> I use JᶻCᶻhᶻ and Jᶻaᶻbᶻ because Jzazbz hurts my eyes trying to look at it;

It's traditionally J<sub>z</sub>C<sub>z</sub>h<sub>z</sub> but if you prefer superscript as opposed to subscript, that's fine. I couldn't tell from your statement if you disliked JzCzhz or disliked  J<sub>z</sub>C<sub>z</sub>h<sub>z</sub> as well. No worries, I won't judge.

> All the "other" rectangular spaces have polar forms (LCH, LCHuv, OKLCh) plus there is HSL, HSV, etc.

I think I found one random reference to LCHuv at one point, but I don't even know if it is common or not, but I used it, because I liked it 🤷🏻.

> (p.s. I use lowercase c for chroma here to differentiate it from the uppercase C which seems to stand for "color" in ICᵀCᴾ as in "color - tritanopia" and "color - protanopia"; 

Often, I see C (chroma) capitalized, but yeah, I get it. I don't know if I've seen in writing the polar form of ICtCp. I don't know what I would specifically call it.

> Thanks again for paving the way for me! I would be a long ways behind in development of my project, if it were not for this one.

No problem. I'm glad you've found it useful in some capacity.

**Reply by SoftMoonWebWare** [reply]
Thanks for your reply.
Either UNICODE or my common Arial font, as best as I can tell (and I looked twice or three times), does not have subscript z t p
I copied and pasted your text above, with  JzCzhz except subscript z, into my text-editor (Komodo Edit), and I got simply JzCzhz.
In fact, I just pasted that here in this message - no subscript.
Unicode may have these characters or not, I don't remember exactly, but I think it did for one; but if I copy it from the Wikipedia page, it shows in my editor as an unrecognized character.
I am now copying Jᶻaᶻbᶻ and ICᵀCᴾ from the text above, and it preserves these "modifier letters" as they are called.
Javascript recognizes them as legal characters in a variable or property or function name.   Same with LChᵤᵥ and HSLᵤᵥ.  (How did I learn that? By reading code in this project, and seeing Greek letters used - they were not legal when I started my project) These subscript and modifier letters are available in most if not all standard fonts, or so it seems so far, anyway.  
So my generic plain-text-output from MasterColorPicker can also output them.  I'm not savvy with these online editors, and I'm not even sure how you got the subscript z to show up (oh, now I see markdown is supported). But the ᵤᵥᶻᵀᴾ characters work with any basic plain-text editor, browser, etc. etc.  Komodo Edit lets me create "snippets" triggered by keystrokes, so I type CTRL+ALT+U, ↓ (arrow-down) and get ᵤ or CTRL+ALT+Z, ↑ (arrow-up) and get ᶻ etc.
Like I said, I go my own way sometimes, and this is one of those times.  It just makes it easier to read.  Nothing against subscript at all!  Peace and equality and respect to all scripts!  But I gotta have something that actually works here....
In my next update, I will have a new "output formatter" dialog where end-users can choose to use the CSS color() function as text-output for color-space-models that CSS supports, and then I drop the fancy characters and make it an eyesore.  Otherwise, end-users and code-readers get the easy-eye version.

Your project is not just useful in SOME capacity.  It is useful in many different capacities.  It does more than my project..in fact it does things that I don't yet even understand or even knew I needed to understand...it teaches me.  But my project is different all together.  Somehow, I never saw "Coloraide" until I saw a link in this project just today.  I haven't looked into the code of that project; but in trying to build my own project from the ground up, looking at other works on GitHub, Wikipedia, etc., I was overwhelmed with questions seemingly without answers, until I found this project.  I gotta admit, I don't know how this Class works internally - that still confuses me; but when it comes to understanding the most important part (to me): color-theory, color conversion algorithms and processes in general, this project is golden.

I only hope my project becomes useful in some capacity.  Maybe it will, because of what I learned here.
Mahalo again and Aloha!

**Reply by LeaVerou** [reply]
If we find a suitable Unicode character, we should change the color space name in the source. That's a nice first-time PR :)

**Reply by SoftMoonWebWare** [reply]
Sorry for the slow response.
For the record, I decided that your perspective, facelessuser, is better.  I implemented IChᵀᴾ (with the capital C for Chroma) in my MasterColorPicker project.  I think it is better to be more consistent.  It makes a nice addition to the BeezEye (color-wheel style) color-picker, alongside LCh, LChᵤᵥ, OKLCh, JᶻCᶻhᶻ, as well as others that conform to sRGB.  I find it amazing how different they all consider "lightness/intensity" to be for various hues.
I'm battling accuracy problems, but I think I got it worked out....needs higher accuracy (more decimal places) in the toString() output.  Hopefully it is a useful tool.  Any other color-pickers out there that do JᶻCᶻhᶻ or IChᵀᴾ?
P.S. that "Evil Russians" LCh/OKLCh color-picker here on GitHub that Google loves (loved) to promote and runs on Node.js seems to activate your computer's camera and change its viewing angle. (I'm intentionally not using their user-name, but a "synonym")  **_read the code before using anything_** Or am I missing something there?  What else would `Camera` be in the Node.js environment? Go to their website (again, I'm not going to mention the address) and it's got stuff running in the background at full-speed; got my CPU fan running at top-speed, when it normally is slow (dragging a map in Google maps will fire it up, also, so it's tied to performance needs).  Initially, I was interested in incorporating their work into the MasterColorPicker. Tried to decipher their website's code, but it is obfuscated; but it seems to use a lot of Web-assembly.  One of those sites that hijack your CPU for some multi-year long computation by 100,000 computers needed to find the trajectory to planet X through wormhole Z?  Not going to find that with the MasterColorPicker.....
Aloha! (and please take Aloha to heart!)

**Reply by facelessuser** [reply]
> P.S. that "Evil Russians" LCh/OKLCh color-picker here on GitHub that Google loves (loved) to promote and runs on Node.js seems to activate your computer's camera and change its viewing angle.

I haven't seen that. My camera does not seem to be activated by their color picker. No camera position is every asked.

Or are you looking through their code? Camera may refer to where to the 3D viewing position is, e.g. the camera viewing the 3D object. Not your real camera.

**Reply by SoftMoonWebWare** [reply]
Yes, I was looking through their code.  I always do before downloading anything.  Haven't been there in a while, but the code used the `Camera` interface, and adjusted the viewing angle.
I'm not sure what you're talking about...a camera viewing a 3D object.  What object to view?  All graphics are 2D in the project I saw online at their website, that had my computer's CPU going full-blast with me just sitting there for several minutes doing nothing.  I don't know a thing about Node.js, but I would assume the `Camera` interface activates the user's camera.  MDN doesn't help.  I've never found **actual** complete instructions for Node.js, just an intro page about creating a server or something, but I use Windows 7, and Node.js does not support it.  I'm praying Apache keeps on keepin' on......
Anyway, if you understand what `Camera` is and does, and why they need it for creating 2D graphics, then I guess you can feel good about using their color-picker.  I don't.  Nor do I feel good about my CPU seemingly being hijacked, like worse than some of those commercial websites throwing ads at you every second, following your mouse-moves, screen-scrolls, etc.  
MasterColorPicker will fire-up your CPU fan if you move your mouse around a lot on a palette with the Color-Space Lab panel open and active - a lot of calculations going on there, along with many HTML inputs (textual and sliders) being changed in real-time.  A lot of overhead.  The CPU fan stops when you stop.

I'm finally getting this latest update finished, with IChᵀᴾ (and ICᵀCᴾ) included, fixing precision errors (in Jᶻaᶻbᶻ and JᶻCᶻhᶻ string-output - not enough), an output formatter, etc.  No time to upload today.

I may pester y'all some more about precision and XYZ conversions and adaptions....later.  For now, thanks again!
Aloha!

**Reply by facelessuser** [reply]
Evil Martian color picker site has 3D models:

<img width="840" alt="Screenshot 2024-11-15 at 3 02 58 PM" src="https://github.com/user-attachments/assets/b9b2492b-983e-40ad-b87b-9e76ebd33c69">

Though, now I'm looking at your post and you say "Russians" not "Martians"? Maybe, I just read what I thought you meant, maybe you are making a joke, maybe you mean something else? I guess if you link to the site you actually are talking about I could comment for sure.

Regardless, if we are talking about "Evil Martians" (this [site](https://oklch.com/#70,0.1,30,100)) my camera light was never turned on, I was never asked for camera permissions, if you search their [code base](https://github.com/search?q=repo%3Aevilmartians%2Foklch-picker%20camera&type=code), "camera" seems to refer to 3D stuff. It's all open. I cannot vouch for the efficiency of everything they are doing as they have multiple dynamic plots. I also don't know the specs of your machine, but mine doesn't seem to go crazy.

You don't have to take my word for it, but it seems pretty clear to me this is talking about 3D cameras.

**Reply by facelessuser** [reply]
It should be noted that they could also be using the same 3D library to render their 2D graphics as well. Nothing prevents you from using a 3D library to generate 2D renders as well.

**Reply by SoftMoonWebWare** [reply]
Yes, "Evil Martians" is what I was referring to.  I didn't want to give them cred and have Google link to them more.  But if you say they're OK, then OK.  Here in the U.S., all we hear about is "Russian" hackers; however I know the U.S. has plenty of them, too, citizens and in the gov'ment.  In plenty other countries, also.  So it was kinda a bad "joke".
If you say that `Camera` has something to do with graphics development, OK.  I really **don't** know.
I've got a Core-I7.  Why mine is "running" and yours is not - IDK.
I would like to know how they do what they do, but I couldn't make heads-nor-tails of what was going on.  Another reason I began to think they were hacking and "using" my machine for their remote computing purpose....whatever that may be.
As far as the `Camera` being turned on with no permissions: the website never asked, and I know a browser will sandbox the website.  But I don't know if Node.js will.  Will it?  Will it ask me every time a .njs file wants to save or read a file locally?  Or open a port to the internet as a server?  I assumed not, the same as PHP on Apache.  It has complete access to your computer, no user-permissions required.
I'm sorry I mentioned it here.  I think it was a bad place for this discussion.  My bad.
Aloha to you, and the "Evil Martians"

**Reply by facelessuser** [reply]
> If you say that Camera has something to do with graphics development, OK. I really don't know.

I am 100% sure that is what they are using "camera" for, I also use "camera" for 3D renders in my project (you can see [here](https://github.com/search?q=repo%3Afacelessuser%2Fcoloraide%20camera&type=code)). IIRC the guy who added 3D models to the Evil Martians color picker had reached out to me and had a lengthy conversation with me on how to render 3D models of color spaces. A snippet from our conversation.

> Hello! I recently reached out about color spaces in coloraide and now my investigations have come to fruition.  I have actually made a pull request to https://oklch.com/ they now have 3d models of lch and oklch color spaces.

Hopefully that puts your mind a little more at ease.

> Here in the U.S.

I am also from the US 🙂.

> But I don't know if Node.js will.

I won't speak to exactly what node.js will or won't do as while I use node, I'm not sure all of what it does. I also can't speak to specific performance concerns of how they implemented what they did in their color picker site. The color picker is all open source though, so you can audit it yourself.
